### PR TITLE
Add `--binning`, `--zoomify` and `--balancing` arguments to pipeline

### PR DESCRIPTION
**Pull request by js2264**
- `--binning INT`: when `--mat_fmt cool`, a **binned** cool matrix can also be generated from pairs file;
- `--zoomify`: if added (and `--binning INT` is present), the binned matrix will be zoomified at `binning` * [1, 2, 5, 10, 20, 50, 100] and balanced
- `--balancing STR` if zoomify, the balancing options for the .mcool file can be specified (cooler default used by default)

So the suggested pipeline call could be something like: 

```{sh}
hicstuff pipeline \
    --enzyme "DpnII,HinfI" \
    --matfmt cool  --binning=1000 --zoomify --balancing "--cis-only" \
    --genome <REF> \
    <R1> <R2>
```

I'm very tempted to change the default `matfmt` to "cool"...

### CODE REVIEW

**Comment by js2264** [comment]
**Note:** I added `cooler` to the `requirements.txt`... I hope this is ok?

**Comment by cmdoret** [comment]
> **Note:** I added `cooler` to the `requirements.txt`... I hope this is ok?

It's currently listed in the [extras_require](https://github.com/koszullab/hicstuff/blob/ef278d2bca7c14fd7402bcc009b7e9fab7f92efa/setup.py#L67) (i.e. optional dependency) since at the time we did not use it as the main format.

I think it would make sense to make it a mandatory dependency, but then it should be removed from the `extras_require`.

File: hicstuff/commands.py [hunk 1]
```diff
@@ -743,8 +743,9 @@ class Pipeline(AbstractCommand):
         pipeline [--aligner=bowtie2] [--centromeres=FILE] [--circular] [--distance-law]
                  [--duplicates] [--enzyme=5000] [--filter] [--force] [--mapping=normal]
                  [--matfmt=graal] [--no-cleanup] [--outdir=DIR] [--plot] [--prefix=PREFIX]
-                 [--quality-min=30] [--read-len=INT] [--remove-centromeres=0] [--size=0]
-                 [--start-stage=fastq] [--threads=1] [--tmpdir=DIR] --genome=FILE <input1> [<input2>]
+                 [--binning=INT] [--zoomify] [--balancing=STR] [--quality-min=30]
```
[the remark this answers is not in the document]
Comment (cmdoret) [reply hicstuff/commands.py:746]: Or maybe even remove enzyme, and only use it in cutsite. What do you think @ABignaud ?

**Review by cmdoret — CHANGES_REQUESTED** [review]
Awesome improvements!! I spotted a mistake and left a few suggestions :)

File: hicstuff/commands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (js2264) [reply hicstuff/commands.py:746]: > Or maybe even remove enzyme, and only use it in cutsite. What do you think @ABignaud ?

We most likely should keep `enzyme`, AFAIU the filtering depends on it.

> If we're adding a "binning" argument, should we disallow integers in `enzyme`?

Also, I believe `enzyme=100` is used for mnase/dnaseC, no? So probably best to still allow integers

**Comment by js2264** [comment]
@cmdoret thanks for your review, I've replied to your comments. Let's wait for @ABignaud input on `emzyme`

File: hicstuff/pipeline.py [hunk 2]
```diff
@@ -339,6 +340,69 @@ def pairs2cool(pairs_file, cool_file, bins_file):
...
+    multires = [res, res*2, res*5, res*10, res*20, res*50, res*100]
+
+    # Run cooler zoomify
+    cooler.zoomify_cooler(clr.filename, output_file, multires, chunksize=10000000)
+
+    # Balance 
+    for res in multires: 
+        cooler_cmd = "cooler balance {args} {cooler_file}::/resolutions/{resolution}"
+        cool_args = {"cooler_file": output_file, "args": balance_args, "resolution": res}
+        sp.call(cooler_cmd.format(**cool_args), shell=True)
```
[the remark this answers is not in the document]
Comment (cmdoret) [reply hicstuff/pipeline.py:402]: I made a mistake: You can actually have spaces in the value, I think the values will just be quoted automatically.

[the remark this answers is not in the document]
Comment (js2264) [reply hicstuff/pipeline.py]: Fixed in [749c895](https://github.com/koszullab/hicstuff/pull/76/commits/749c895c1f2f2de093293cd47bfc2597156b4967)

File: hicstuff/commands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (js2264) [reply hicstuff/commands.py]: Ok we'll keep allowing integers for `enzyme` for now

**Pull request merged**